### DISCUSSION TITLE
纯后端校验schema数据能力，不依赖表单UI

### DISCUSSION THREAD

**Discussion by Jiazhi-Wong** [alibaba/formily#1797] [Show and tell]
## 目的

目前formily的schema校验能力是依赖表单的，前端可以在保存数据前校验拦截保存操作。

但是如果校验的业务场景是没有UI表单的情况，比如：
- 页面有多个表单，表单组件会被销毁，在最后保存提交的时候再做所有表单的校验
- 在node服务端的后端场景下，对数据进行二次校验

总而言之，我们需要一个不依赖UI的校验 schema 数据的能力，而且校验规则要跟 formily 表单保持一致。

## 思路

研究源码发现，解析 schema 并生成 form、field 的能力是在 @formily/react 完成的，所以依赖UI层的能力。

而我们的解决思路是直接对接 @formily/core 和 @formily/json-schema，递归 schema 结构，直接生成对应的 form、field 模型。

感谢 formily 的分层设计合理，@formily/core 是完全不依赖UI的，并且校验模型也是不依赖UI，校验能力也是前后端一致，才得以实现这个功能。

## demo

https://codesandbox.io/s/vibrant-thompson-bgoc2?file=/index.ts

demo代码不一定考虑周全，麻烦 @janryWang 大大看看有没有漏掉什么细节

**Comment by janryWang** [comment]
第一次发现还能这么用，感觉很酷啊，相当于是把core作为一个虚拟表单，只用于做异步校验

**Reply by Jiazhi-Wong** [reply]
有这样的业务需求，我研究了源码，才想到的方法。还是 formily 整体设计上比较合理，所以无论需求是什么，都有办法接入。

**Reply by TanYJie** [reply]
这个 demo 确实很有酷很有用。其实我觉得类似 designable 这种搭建系统，也会有这种需求。校验页面中所有组件的配置项时，需要对每个节点调用这种 validateBySchema 的方法，而且还要考虑联动等情况。希望 formily 将来可以官方提供这个方法

**Comment by janryWang** [comment]
感觉可以再封装一下，API就叫做validateBySchema，同时注意一下，在校验结束的时候调用一下form.onUnmount释放内存，否则会内存泄漏的

```

validateBySchema(values,schema).then(()=>{})

```

**Comment by darkmice** [comment]
可以用一个微服务专干这种事，😁

**Comment by atzcl** [comment]
我也做过类似的需求，不过用的是 ajv，现在想想，完全可以复用 formily 自带的 json schema 能力，还省事了

**Comment by Jiazhi-Wong** [comment]
@janryWang 大大，今天升级了 formily，从 2.0.0-beta.78 => 2.0.0-rc.11，这个 validateBySchema 检查不出错误了，请问是改动了？
2.0.0-beta.78 demo：https://codesandbox.io/s/vibrant-thompson-bgoc2
2.0.0-rc.11 demo：https://codesandbox.io/s/snowy-bash-b9beo

现在我想升级到 2.0.0-rc.11，也保留 validateBySchema 的能力，应该怎么做？

**Comment by janryWang** [comment]
是个bug，麻烦提个issue

**Comment by Jiazhi-Wong** [comment]
@janryWang 大大，还有个新问题，如果 schema 是 required 并且有 default 值，就不能校验到 `undefined` 值，比如下面这个demo：
https://codesandbox.io/s/cool-hooks-5ewer?file=/index.ts

期望是 `stringWithDefault` 字段也能校验出报错，应该怎么做？

**Reply by janryWang** [reply]
合理的，有默认值的时候，字段初始值就是默认值，校验合法

**Reply by Jiazhi-Wong** [reply]
已经找到解决办法了，在校验前，把 `default` 值删掉就好了

**Comment by eric-gitta-moore** [comment]
考古+1，同样场景，只需要校验和reaction功能，但是要和框架响应式绑定上。
场景是 server -> mitm(注入后端没有的字段，以及类似 computed 字段) -> vue -> UI(用户填写部分) -> merge schema -> submit -> validate -> server

**Reply by eric-gitta-moore** [reply]
对于自带响应式的框架，其实感觉还可以在智能一点。比如 vue，把reactive 之后的数据再包一层 observer，然后设置为 form 的 values。这样 formily 写到自己数据的同时也会写入到 vue reactive 里面，这样用 vue 自带的响应式就能完成各种操作了。

```tsx
import { reactive, isReactive } from 'vue';
import { observable, isObservable } from '@formily/reactive';

// 这是需要校验的数据
const schemaData = observable(reactive({
  stringWithDefault: undefined,
  stringWithoutDefault: undefined
}));

console.log(isReactive, isReactive(schemaData)) // true
console.log(isObservable, isObservable(schemaData)) // true

const form = createForm({
  values: schemaData 
});
```
